### ISSUE TITLE
[LSP] Words surrounded by back quotes should be blue in markdown comments

### ISSUE THREAD

**Issue by acoutts** [Dart-Code/Dart-Code#3330]
**Describe the bug**
Previously if you surrounded a word with back quotes it would turn blue in a markdown comment (triple slash), but it no longer does with LSP.

**To Reproduce**
Add this comment:
```
/// `this should be blue`
```

**Expected behavior**
Back quotes should emphasize things as before.

**Screenshots**
<img width="411" alt="Screen Shot 2021-05-08 at 9 01 20 AM" src="https://user-images.githubusercontent.com/9597207/117540243-4332ca00-afdc-11eb-825d-0742734e7557.png">

**Versions (please complete the following information):**
 - VS Code version:
```
1.57.0-insider
Commit: b95b73b7b074f5dddccd57e5e290617f1a767ec1
```
 - Dart extension version: `3.22.0`
 - Dart/Flutter SDK version:
```
Flutter 2.3.0-0.1.pre • channel unknown • unknown source
Framework • revision d72bfb8d07 (12 days ago) • 2021-04-26 06:05:55 -0700
Engine • revision de6e1adf97
Tools • Dart 2.14.0 (build 2.14.0-18.0.dev)
```

**Comment by DanTup** [comment]
@bwilkerson this used to work because of regex in the textmate grammar, though the semantic tokens (which only cover the whole comment) take precedence.

Some possible thoughts:

- Don't send comment blocks and leave it to the textmate grammar (this isn't good for non-VS Code clients though)
- Parse the comments as markdown to identify code blocks and add a modifier to them
- Use a simple regex to identify simple expressions on a single line in backticks (this wouldn't support any over multiple lines, or other types of code blocks like triple-backtick blocks)

None of these feel like an ideal solution, but it would be a shame to lose this highlighting. Do you have any thoughts on what might be best to do?

**Comment by bwilkerson** [comment]
There has been some discussion before of using the `markdown` parser to improve our handling of documentation comments, but I haven't looked at how difficult that might be to do. But that would certainly be my preference. Not only could we improve semantic highlighting, but we could also improve the display of comments in hovers and code completion.

I've seen simple regex style parsing used in some editors and it tends to break down fairly easily, so I'd be hesitant to use it here. I'm not sure it wouldn't be worse than not having any highlighting.

**Comment by DanTup** [comment]
Parsing them could also make it easier to re-wrap them (something I've wanted to do for a while, but ensuring it doesn't mess with embedded code/lists/etc without parsing also seems error-prone).

I had a quick look at the [markdown package](https://github.com/dart-lang/markdown) to see if this might be easy, but if I'm [understanding this code correctly](https://github.com/dart-lang/markdown/blob/master/lib/src/block_parser.dart#L474-L489) it doesn't provide offsets/lengths when producing nodes - if it did, perhaps it would be simple to have a visitor that just collects the offsets/lengths of any `code` elements to handle the above.

**Comment by bwilkerson** [comment]
We could probably extend the parser to keep the extra information we need. I doubt anyone would object, but we could create an issue in that repo proposing the changes and see whether anyone knows of a reason why it wouldn't be acceptable. If there's a problem then we could employ a method similar to what we did with the YAML parser and add an API to return richer objects.

**Comment by DanTup** [comment]
sgtm - I've filed https://github.com/dart-lang/markdown/issues/369 to start a discussion.

**Comment by jducaud** [comment]
This behavior concerns all the highlighted items in doc comments, meaning not only items surrounded in back quotes, but also references surrounded in square brackets and code blocks surrounded in triple back quotes.

**Dart Code extension disabled:**
![without Dart extension](https://user-images.githubusercontent.com/55034588/121685647-79bea180-cac0-11eb-9f70-46f55782cb7f.png)

**Dart Code extension enabled:**
![with dart extension](https://user-images.githubusercontent.com/55034588/121685710-90fd8f00-cac0-11eb-97e6-e3506abc271e.png)

**Comment by DanTup** [comment]
@jducaud yep, understood. The difference between your screenshots is really the difference between LSP semantic tokens or the textmate grammar. The reason it works without the extension enabled now is that in the recent release of VS Code includes the textmate grammar from this extension directly in VS Code.

If you want to disable semantic tokens for now until this is addressed, you can add the following to your VS Code user settings:

```ts
"[dart]": {
	"editor.semanticHighlighting.enabled": false
}
```

This will result in the same colouring as your first screenshot without needing to disable LSP or the extension. It does mean a few things may not be coloured correctly due to limitations of the regular expressions (for example if you name a method `set` it'll be coloured like the `set` keyword instead of a method), though you might it's worth the trade-off.

**Comment by jducaud** [comment]
Thank you @DanTup! This trick is a good provisional workaround
